core/kernel/shared/sprintf.rb

```diff
@@ -0,0 +1,945 @@
+        format("%1$*2$a", 196, 20).should == "           0x1.88p+7"
+        format("%1$*2$A", 196, 20).should == "           0X1.88P+7"
+
+        format("%1$*2$c", 97, 10).should == "         a"
+        format("%1$*2$p", [], 10).should == "        []"
+        format("%1$*2$s", "abc", 10).should == "       abc"
+      end
+
+      it "raises ArgumentError when is mixed with width" do
+        -> () {
+          format("%*10d", 10, 112)
+        }.should raise_error(ArgumentError)
+      end
+    end
+  end
+
+  describe "width" do
+    it "specifies the minimum number of characters that will be written to the result" do
+      format("%10b", 10).should == "      1010"
+      format("%10B", 10).should == "      1010"
+      format("%10d", 112).should == "       112"
+      format("%10i", 112).should == "       112"
+      format("%10o", 87).should == "       127"
+      format("%10u", 112).should == "       112"
+      format("%10x", 196).should == "        c4"
+      format("%10X", 196).should == "        C4"
+
+      format("%20e", 109.52).should == "        1.095200e+02"
+      format("%20E", 109.52).should == "        1.095200E+02"
+      format("%20f", 10.952).should == "           10.952000"
+      format("%20g", 12.1234).should == "             12.1234"
+      format("%20G", 12.1234).should == "             12.1234"
+      format("%20a", 196).should == "           0x1.88p+7"
+      format("%20A", 196).should == "           0X1.88P+7"
+
+      format("%10c", 97).should == "         a"
+      format("%10p", []).should == "        []"
+      format("%10s", "abc").should == "       abc"
+    end
+
+    it "is ignored if argument's actual length is greater" do
+      format("%5d", 1234567890).should == "1234567890"
+    end
+  end
+
+  describe "precision" do
+    context "integer types" do
+      it "controls the number of decimal places displayed" do
+        format("%.6b", 10).should == "001010"
+        format("%.6B", 10).should == "001010"
+        format("%.5d", 112).should == "00112"
+        format("%.5i", 112).should == "00112"
+        format("%.5o", 87).should == "00127"
+        format("%.5u", 112).should == "00112"
+
+        format("%.5x", 196).should == "000c4"
+        format("%.5X", 196).should == "000C4"
+      end
+    end
+
+    context "float types" do
+      it "controls the number of decimal places displayed in fraction part" do
+        format("%.10e", 109.52).should == "1.0952000000e+02"
+        format("%.10E", 109.52).should == "1.0952000000E+02"
+        format("%.10f", 10.952).should == "10.9520000000"
+        format("%.10a", 196).should == "0x1.8800000000p+7"
+        format("%.10A", 196).should == "0X1.8800000000P+7"
+      end
+
+      # strange behavior
+      it "does not affect G format" do
+        format("%.10g", 12.1234).should == "12.1234" # expected "12.1234000000"
+        format("%.10g", 123456789).should == "123456789" # expected "1.2345700000e+08"
```

Comment:
Sounds like a bug, could you report it? (I think a single issue asking multiple strange behaviors for sprintf is OK)

Reply:
First case: is ok. `g` does not print trailing zeroes, but there's the `#` flag if you want that.
Second case: also ok. Precision is 10, the integer has 9 digits so fits without resorting to scientific notation. Use two more digits, or two less precision, to get scientific notation.

Doc says:
> Convert a floating point number using exponential form
> if the exponent is [...] or greater than or
> equal to the precision, [...]

Reply:
Thanks for the clarification.